zomes/likes/code/src/lib.rs

```diff
@@ -0,0 +1,184 @@
+        // This is super important to make the DHT Sybil resistant
+        // even though the app itself is permissionless (as is the case here).
+        //
+        // In this example we need to restrict the DHT with the same
+        // custom "like_membrane" (defined below) because that membrane
+        // uses the DHT for its implementation.
+        // Would every node be inlcuded in the DHT (and trusted), attackers
+        // could easily fool a node into believing their constructed
+        // like-reality.
+        hc_dht_membrane: |object: MembraneObject| {
+            hdk::check_membrane("liked", object.agent_key())
+        },
+
+        // This is a custom membrane that we can use in zome functions
+        // and in entry validations.
+        // (My intention is to let only "liked" people comment on posts).
+        //
+        // This is basically an implementation of social triangulation.
+        // You need a certain number of people who are "liked" to also
+        // like you.
+        //
+        // Of course, this only works when we now with whom to start.
+        // Those seed agents as well as the number of needed likes
+        // are properties, i.e. defined in and thus part of the DNA.
+        liked_membrane: |object: MembraneObject| {
+            let agent = object.agent_key();
+            // Get the seed agents that are defined in the DNA
+            let seeds : Vec<String> = hdk::properties("seeds");
+
+            // Seeds are liked by definition
+            if seeds.include(agent) {
+                true
+            } else {
+                 // I'm using a LikedNotarization entry to constrain
+                 // the transitive requirement of this membrane
+                 // (that only likes of liked ones are valid).
+                 // We don't want to run a depth-search on the like-graph
+                 // every time we have to check the membrane - instead
+                 // we use the trusted DHT as cache:
+                 if hdk::get_links(agent, "like_notarization").size() > 0 {
+                    true
+                } else {
+                    // Ok, no seed agent and no cached notarization yet.
+                    // We have to check ourselves:
+                    let likes = hdk::get_links(agent, "received_likes");
+                    let min_likes = hdk::properties("required_number_of_likes");
+                    // if the number of likes is less than need, don't bother..
+                    if likes.size() >= min_likes {
+                        // But if we have enough likes we need to check
+                        // how many of them are actually valid, i.e. coming
+                        // from a liked source
+                        let num_valid_likes = likes.iter()
+                            .filter(|like| {
+                                let like_source = hdk::get_links(like, "source")[0];
+                                hdk::check_membrane(like_source)
```

Comment:
So... what is it that the proposed `hdk::check_membrane` actually does under the hood?

Reply:
It runs the callbacks (or the built-in code if it is a default/system membrane) needed to check if the given agent is in the given membrane. (yes, the latter is missing here, will change the line. It should be `hdk::check_membrane("liked", like_source)`). So this is actually a recursive call of the same function but with source of the likes of the first agent..